src/QuadVelocityController.cpp

```diff
@@ -122,19 +128,46 @@ bool QuadVelocityController::getVelocities(geometry_msgs::Vector3& return_veloci
                 velocities_valid = false;
             }
 
-            // Get the transforms with the stamps for readability
+            // Get the transforms without the stamps for readability
             geometry_msgs::Transform& transform = transformStamped.transform;
             geometry_msgs::Transform& oldTransform = last_transform_stamped_.transform;
 
+            // Get the yaw (z axis) rotation from the quanternion
+            double ysqr = transform.rotation.y * transform.rotation.y;
+            double t3 = 2.0f * (transform.rotation.w * transform.rotation.z + transform.rotation.x * transform.rotation.y);
+            double t4 = 1.0f - 2.0f * (ysqr + transform.rotation.z * transform.rotation.z);  
+            current_yaw = std::atan2(t3, t4);
+
             // Calculate x, y, and z velocity
+            // X and Y are transformed using the current yaw heading to velocities that the pitch and roll can respond to directly
             double delta = delta_seconds.toSec();
+            double world_x_vel = ((transform.translation.x - oldTransform.translation.x) / delta);
+            double world_y_vel = ((transform.translation.y - oldTransform.translation.y) / delta);
+            return_velocities.linear.x = world_x_vel * std::cos(current_yaw) + world_y_vel * std::sin(current_yaw);
+            return_velocities.linear.y = world_x_vel * -std::sin(current_yaw) + world_y_vel * std::cos(current_yaw);
+
+            return_velocities.linear.z = (transform.translation.z - oldTransform.translation.z) / delta;
+
+            double yaw_difference = current_yaw - last_yaw_;
+
+            // The next two if statements handle if the yaw_difference is large due to gimbal lock
+            // Assumes that we don't have a yaw_difference more than 2pi. We won't since we get all the angles
+            // from an atan2 which only return -pi to pi
+            // Also assumes that we won't have a rotation difference more than pi
+            if(yaw_difference > 3.14159)
```

Comment:
C++ doesn't have a built-in constant in the spec: http://stackoverflow.com/questions/1727881/how-to-use-the-pi-constant-in-c

It's also defined in boost.

However it's available in gcc easily these day's it's just not standard.